seatunnel-translation/seatunnel-translation-flink/seatunnel-translation-flink-13/src/main/java/org/apache/seatunnel/translation/flink/metric/FlinkMetricContext.java

```diff
@@ -0,0 +1,120 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Field;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class FlinkMetricContext implements MetricsContext {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FlinkMetricContext.class);
+
+    private final Map<String, Metric> metrics = new ConcurrentHashMap<>();
+
+    private MetricGroup metricGroup;
+
+    private StreamingRuntimeContext runtimeContext;
+
+    public FlinkMetricContext(MetricGroup metricGroup) {
+        this.metricGroup = metricGroup;
+    }
+
+    public FlinkMetricContext(StreamingRuntimeContext runtimeContext) {
+        this.runtimeContext = runtimeContext;
+    }
+
+    @Override
+    public Counter counter(String name) {
+        if (metrics.containsKey(name)) {
+            return (Counter) metrics.get(name);
+        }
+        Counter counter =
+                runtimeContext == null
+                        ? new FlinkGroupCounter(name, metricGroup.counter(name))
+                        : new FlinkCounter(name, runtimeContext.getLongCounter(name));
+        return this.counter(name, counter);
+    }
+
+    @Override
+    public <C extends Counter> C counter(String name, C counter) {
+        this.addMetric(name, counter);
+        return counter;
+    }
+
+    @Override
+    public Meter meter(String name) {
+        if (metrics.containsKey(name)) {
+            return (Meter) metrics.get(name);
+        }
+
+        // Why use reflection to obtain metrics group?
+        // Because the value types returned by flink 1.13 and 1.14 runtimeContext.getMetricGroup()
+        // are inconsistent
+        org.apache.flink.metrics.Meter meter;
+        if (runtimeContext == null) {
+            meter = metricGroup.meter(name, new MeterView(5));
+        } else {
+            try {
+                Field field = AbstractRuntimeUDFContext.class.getDeclaredField("metrics");
+                field.setAccessible(true);
+                MetricGroup mg = (MetricGroup) field.get(runtimeContext);
+                meter = mg.meter(name, new MeterView(5));
+            } catch (Exception e) {
+                throw new IllegalStateException("Initial meter failed", e);
+            }
+        }
+        return this.meter(name, new FlinkMeter(name, meter));
+    }
+
+    @Override
+    public <M extends Meter> M meter(String name, M meter) {
+        this.addMetric(name, meter);
+        return meter;
+    }
+
+    protected void addMetric(String name, Metric metric) {
+        if (metric == null) {
+            LOGGER.warn("Ignoring attempted add of a metric due to being null for name {}.", name);
+        } else {
+            synchronized (this) {
+                Metric prior = this.metrics.put(name, metric);
+                if (prior != null) {
+                    this.metrics.put(name, prior);
+                    LOGGER.warn(
```

Comment:
Should an exception be thrown?

Reply:
I drew inspiration from the addition logic in `AbstractMetricsContext`. In the original design, if duplicate metrics were added, the later ones would not take effect. I am not sure about the design details, @ic4y please confirm the details. In my personal opinion, it is okay not to throw any exceptions here.

Reply:
sure, Either throw or log.warn works